dev/TabView/TabView.cpp

```diff
@@ -0,0 +1,248 @@
+            {
+                scrollContentPresenter.TabFocusNavigation(winrt::KeyboardNavigationMode::Once);
+            }
+        }
+    }
+
+    UpdateTabWidths();
+}
+
+void TabView::OnSizeChanged(const winrt::IInspectable& sender, const winrt::SizeChangedEventArgs& args)
+{
+    UpdateTabWidths();
+}
+
+void TabView::OnItemsChanged(winrt::IInspectable const& item)
+{
+    if (auto args = item.as< winrt::IVectorChangedEventArgs>())
+    {
+        int numItems = (int)Items().Size();
+        if (args.CollectionChange() == winrt::CollectionChange::ItemRemoved && numItems > 0)
+        {
+            if (SelectedIndex() == (int32_t)args.Index())
+            {
+                // Find the closest tab to select instead.
+                int startIndex = (int)args.Index();
+                if (startIndex >= numItems)
+                {
+                    startIndex = numItems - 1;
+                }
+                int index = startIndex;
+
+                do
+                {
+                    auto nextItem = ContainerFromIndex(index).as<winrt::ListViewItem>();
+
+                    if (nextItem && nextItem.IsEnabled() && nextItem.Visibility() == winrt::Visibility::Visible)
+                    {
+                        // We need to wait until OnSelectionChanged fires to change the selection, otherwise it will get lost.
+                        m_indexToSelectOnSelectionChanged = std::optional(index);
+                        break;
+                    }
+
+                    // try the next item
+                    index++;
+                    if (index >= numItems)
+                    {
+                        index = 0;
+                    }
+                } while (index != startIndex);
+            }
+        }
+    }
+
+    UpdateTabWidths();
+
+    __super::OnItemsChanged(item);
+}
+
+void TabView::OnSelectionChanged(const winrt::IInspectable& sender, const winrt::SelectionChangedEventArgs& args)
+{
+    if (m_indexToSelectOnSelectionChanged)
+    {
+        SelectedItem(Items().GetAt(m_indexToSelectOnSelectionChanged.value()));
+        m_indexToSelectOnSelectionChanged = {};
+    }
+
+    UpdateTabContent();
+}
+
+void TabView::UpdateTabContent()
+{
+    if (auto tabContentPresenter = m_tabContentPresenter.get())
+    {
+        if (!SelectedItem())
+        {
+            tabContentPresenter.Content(nullptr);
+            tabContentPresenter.ContentTemplate(nullptr);
+        }
+        else
+        {
+            auto container = ContainerFromItem(SelectedItem()).as<winrt::ListViewItem>();
+            if (container)
+            {
+                tabContentPresenter.Content(container.Content());
+                tabContentPresenter.ContentTemplate(container.ContentTemplate());
+            }
+        }
+    }
+}
+
+void TabView::CloseTab(winrt::TabViewItem const& container)
+{
+    if (auto item = ItemFromContainer(container))
```

Comment:
>ItemFromContainer [](start = 20, length = 17)

It's more efficent to use IndexFromContainer?

Reply:
If I only needed the index, that's true, but I need the item too.